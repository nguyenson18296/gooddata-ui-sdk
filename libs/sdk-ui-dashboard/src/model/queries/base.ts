// (C) 2021 GoodData Corporation

/**
 * @internal
 */
export type DashboardQueryType = "GDC.DASH/QUERY.INSIGHT.DATE.DATASETS";

/**
 * Base type for all dashboard queries. A dashboard query encapsulates how complex, read-only dashboard-specific logic
 * can be can be executed.
 *
 * @internal
 */
export interface IDashboardQuery<_TResult = any> {
    /**
     * Query type. Always starts with "GDC.DASH/QUERY".
     */
    readonly type: DashboardQueryType;

    /**
     * Correlation ID can be provided when creating a query. Events emitted during the query processing
     * will contain the same correlation ID.
     */
    readonly correlationId?: string;
}
