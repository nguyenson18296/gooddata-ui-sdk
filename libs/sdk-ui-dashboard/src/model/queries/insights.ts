// (C) 2021 GoodData Corporation
import { IDashboardQuery } from "./base";
import { ObjRef } from "@gooddata/sdk-model";

/**
 * Given a reference to an insight, this query will obtain list of all date datasets that may be used
 * to filter it.
 *
 * @internal
 */
export interface QueryDateDatasetsForInsight extends IDashboardQuery<DateDatasetsForInsight> {
    type: "GDC.DASH/QUERY.INSIGHT.DATE.DATASETS";
    payload: {
        insightRef: ObjRef;
    };
}

/**
 * Creates action through which you can query dashboard component for information about the date data sets
 * that are applicable for the provided insight.
 *
 * @param insightRef - reference to insight
 * @internal
 */
export function queryDateDatasetsForInsight(
    insightRef: ObjRef,
    correlationId?: string,
): QueryDateDatasetsForInsight {
    return {
        type: "GDC.DASH/QUERY.INSIGHT.DATE.DATASETS",
        correlationId,
        payload: {
            insightRef,
        },
    };
}

/**
 * @internal
 */
export type DateDatasetsForInsight = {
    data: any;
};
