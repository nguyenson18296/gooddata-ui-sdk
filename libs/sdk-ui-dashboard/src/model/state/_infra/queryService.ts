// (C) 2021 GoodData Corporation
import { DashboardContext } from "../../types/commonTypes";
import { SagaIterator } from "redux-saga";
import {
    CaseReducer,
    createEntityAdapter,
    createSelector,
    createSlice,
    EntityId,
    EntityState,
    IdSelector,
    OutputSelector,
    PayloadAction,
    Slice,
} from "@reduxjs/toolkit";
import { DashboardQueryType, IDashboardQuery } from "../../queries/base";
import { DashboardState } from "../types";
import { CaseReducerActions } from "@reduxjs/toolkit/src/createSlice";
import { call, put, SagaReturnType, select } from "redux-saga/effects";

/**
 *
 */
export interface IDashboardQueryService<TQuery extends IDashboardQuery, TResult> {
    name: DashboardQueryType;
    generator: (ctx: DashboardContext, query: TQuery) => SagaIterator<TResult>;
    cache: QueryCache<TQuery, TResult>;
}

type QueryCacheEntry<TQuery extends IDashboardQuery, TResult> = {
    query: TQuery;
    result: TResult;
};

type QueryCacheReducer<TQuery extends IDashboardQuery, TResult, TPayload> = CaseReducer<
    EntityState<QueryCacheEntry<TQuery, TResult>>,
    PayloadAction<TPayload>
>;

type AllQueryCacheReducers<TQuery extends IDashboardQuery, TResult> = {
    set: QueryCacheReducer<TQuery, TResult, QueryCacheEntry<TQuery, TResult>>;
    remove: QueryCacheReducer<TQuery, TResult, string>;
    removeAll: QueryCacheReducer<TQuery, TResult, void>;
};

type QueryActions<TQuery extends IDashboardQuery, TResult> = CaseReducerActions<
    AllQueryCacheReducers<TQuery, TResult>
>;

type QueryCache<TQuery extends IDashboardQuery, TResult> = {
    slice: Slice<
        EntityState<QueryCacheEntry<TQuery, TResult>>,
        AllQueryCacheReducers<TQuery, TResult>,
        string
    >;
    actions: QueryActions<TQuery, TResult>;
    selectById: (
        id: EntityId,
    ) => OutputSelector<DashboardState, TResult | undefined, (res: DashboardState) => TResult | undefined>;
};

function createQueryCacheSlice<TQuery extends IDashboardQuery, TResult>(
    queryName: string,
    selectId: IdSelector<TQuery>,
): QueryCache<TQuery, TResult> {
    const cacheEntryId = (entry: QueryCacheEntry<TQuery, TResult>) => selectId(entry.query);
    const adapter = createEntityAdapter({ selectId: cacheEntryId });
    // TODO: getting massive typing error here, don't know how to solve it
    const reducers: AllQueryCacheReducers<TQuery, TResult> = {
        set: adapter.addOne as any,
        remove: adapter.removeOne as any,
        removeAll: adapter.removeAll as any,
    };

    const slice = createSlice<
        EntityState<QueryCacheEntry<TQuery, TResult>>,
        AllQueryCacheReducers<TQuery, TResult>
    >({
        name: queryName,
        initialState: adapter.getInitialState(),
        reducers,
    });

    const selectors = adapter.getSelectors((state: DashboardState) => state._queryCache[queryName]);
    const selectById = (id: EntityId) => {
        return createSelector(
            (state: DashboardState) => state,
            (state) => {
                return selectors.selectById(state, id)?.result;
            },
        );
    };

    return {
        slice,
        actions: slice.actions,
        selectById,
    };
}

export type QueryServiceCachingConfig<TQuery extends IDashboardQuery, TResult> = {
    queryName: DashboardQueryType;
    generator: (ctx: DashboardContext, query: TQuery) => SagaIterator<TResult>;
    queryToCacheKey: (query: TQuery) => EntityId;
};

export function createCachedQueryService<TQuery extends IDashboardQuery, TResult>(
    config: QueryServiceCachingConfig<TQuery, TResult>,
): IDashboardQueryService<TQuery, TResult> {
    const { queryName, queryToCacheKey, generator } = config;
    const queryCache = createQueryCacheSlice<TQuery, TResult>(queryName, queryToCacheKey);

    function* generatorWithQueryCache(ctx: DashboardContext, query: TQuery): SagaIterator<TResult> {
        const cacheKey = queryToCacheKey(query);
        const cachedResult = yield select(queryCache.selectById(cacheKey));

        if (cachedResult) {
            return cachedResult;
        }

        const result: SagaReturnType<typeof generator> = yield call(generator, ctx, query);

        yield put(
            queryCache.actions.set({
                query,
                result,
            }),
        );

        return result;
    }

    return {
        name: queryName,
        generator: generatorWithQueryCache,
        cache: queryCache,
    };
}
