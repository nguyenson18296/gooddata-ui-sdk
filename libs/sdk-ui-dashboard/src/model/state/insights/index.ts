// (C) 2021 GoodData Corporation

import { createSlice } from "@reduxjs/toolkit";
import { insightsAdapter } from "./insightsEntityAdapter";

const insightsSlice = createSlice({
    name: "insights",
    initialState: insightsAdapter.getInitialState(),
    reducers: {
        setInsights: insightsAdapter.setAll,
        addInsights: insightsAdapter.addMany,
    },
});

export const insightsSliceReducer = insightsSlice.reducer;
export const insightsActions = insightsSlice.actions;
