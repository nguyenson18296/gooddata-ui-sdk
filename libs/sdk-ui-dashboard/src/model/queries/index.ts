// (C) 2021 GoodData Corporation

import { QueryDateDatasetsForInsight } from "./insights";

export { IDashboardQuery, DashboardQueryType } from "./base";
export { QueryDateDatasetsForInsight, DateDatasetsForInsight, queryDateDatasetsForInsight } from "./insights";

/**
 * @internal
 */
export type DashboardQueries = QueryDateDatasetsForInsight;
