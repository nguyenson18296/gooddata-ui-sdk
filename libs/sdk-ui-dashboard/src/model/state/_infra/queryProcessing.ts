// (C) 2021 GoodData Corporation

import { IDashboardQueryService } from "./queryService";
import { Saga, SagaIterator } from "redux-saga";
import { actionChannel, call, getContext, spawn, take } from "redux-saga/effects";
import { IDashboardQuery } from "../../queries";
import { DashboardContext } from "../../types/commonTypes";
import keyBy from "lodash/keyBy";
import { Action, CombinedState, combineReducers, Reducer } from "@reduxjs/toolkit";
import fromPairs from "lodash/fromPairs";
import noop from "lodash/noop";
import compact from "lodash/compact";
import { dispatchDashboardEvent } from "../../eventEmitter/eventDispatcher";
import {
    internalQueryErrorOccurred,
    isDashboardQueryFailed,
    queryCompleted,
    queryRejected,
    queryStarted,
} from "../../events/general";

/**
 * Query processing component has multiple pieces that need to be integrated into the redux store.
 */
export type QueryProcessingModule = {
    /**
     * Query services may store the results in state for caching purposes. All services that use caching implement
     * the cache as a separate slice of the internal `_queryCache` part of the state. This reducer is a combined
     * reducer including all the appropriate slice reducers.
     */
    queryCacheReducer: Reducer<CombinedState<any>>;

    /**
     * A single saga is in place to handle query processing requests. Query requests will be processed concurrently.
     */
    rootQueryProcessor: Saga;
};

/**
 * @internal
 */
export const QueryEnvelopeActionTypeName = "@@QUERY.ENVELOPE";

/**
 * @internal
 */
export type QueryEnvelope = {
    readonly type: typeof QueryEnvelopeActionTypeName;
    readonly query: IDashboardQuery;
    readonly onStart: (query: any) => void;
    readonly onSuccess: (result: any) => void;
    readonly onError: (err: Error) => void;
};

function* processQuery(
    service: IDashboardQueryService<any, any>,
    ctx: DashboardContext,
    envelope: QueryEnvelope,
) {
    const {
        query,
        query: { type, correlationId },
    } = envelope;
    const correlationIdForLog = correlationId ?? "(no correlationId provided)";

    try {
        try {
            envelope.onStart(query);
        } catch (e) {
            // eslint-disable-next-line no-console
            console.warn(
                `An error has occurred while calling onStart function provided for ${type}@${correlationIdForLog} processing:`,
                e,
            );
        }

        yield dispatchDashboardEvent(queryStarted(ctx, correlationId));

        const result: SagaIterator<typeof service.generator> = yield call(
            service.generator,
            ctx,
            envelope.query,
        );

        try {
            envelope.onSuccess(result);
        } catch (e) {
            // eslint-disable-next-line no-console
            console.warn(
                `An error has occurred while calling onSuccess function provided for ${type}@${correlationIdForLog} processing`,
                e,
            );
        }
        yield dispatchDashboardEvent(queryCompleted(ctx, query, result, correlationId));
    } catch (e) {
        try {
            envelope.onError(e);
        } catch (ne) {
            // eslint-disable-next-line no-console
            console.warn(
                `An error has occurred while calling onError function provided for ${type}@${correlationIdForLog} processing:`,
                ne,
            );
        }

        if (isDashboardQueryFailed(e)) {
            yield dispatchDashboardEvent(e);
        } else {
            yield dispatchDashboardEvent(
                internalQueryErrorOccurred(
                    ctx,
                    `An internal error has occurred while processing ${type}`,
                    e,
                    correlationId,
                ),
            );
        }
    }
}

function ensureQueryWrappedInEnvelope(action: Action): QueryEnvelope {
    if (action.type === QueryEnvelopeActionTypeName) {
        return action as QueryEnvelope;
    }

    return {
        type: QueryEnvelopeActionTypeName,
        query: action,
        onStart: noop,
        onSuccess: noop,
        onError: noop,
    };
}

/**
 * Creates components that should be integrated into the dashboard store in order to facilitate query processing.
 *
 * @param queryServices - query services use to initialize the components
 */
export function createQueryProcessingModule(
    queryServices: IDashboardQueryService<any, any>[],
): QueryProcessingModule {
    const servicesByType = keyBy(queryServices, (service) => service.name);
    const queryToReducers = fromPairs(
        compact(
            queryServices.map((service) => {
                if (!service.cache) {
                    return null;
                }

                return [service.cache.cacheName, service.cache.reducer];
            }),
        ),
    );

    return {
        queryCacheReducer: combineReducers(queryToReducers),
        /*
         * The root saga for all query processing. This will channel in all query envelopes and all non-enveloped
         * queries and will dispatch the query
         */
        rootQueryProcessor: function* (): SagaIterator<void> {
            const queryChannel = yield actionChannel(
                (action: any) =>
                    action.type === QueryEnvelopeActionTypeName || action.type.startsWith("GDC.DASH/QUERY."),
            );

            while (true) {
                const query = yield take(queryChannel);
                const envelope = ensureQueryWrappedInEnvelope(query);
                const dashboardContext: DashboardContext = yield getContext("dashboardContext");
                const service = servicesByType[envelope.query.type];

                if (!service) {
                    yield dispatchDashboardEvent(
                        queryRejected(dashboardContext, envelope.query.correlationId),
                    );
                } else {
                    yield spawn(processQuery, service, dashboardContext, envelope);
                }
            }
        },
    };
}
