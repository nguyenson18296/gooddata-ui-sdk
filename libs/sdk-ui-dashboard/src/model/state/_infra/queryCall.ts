// (C) 2021 GoodData Corporation

import { IDashboardQuery } from "../../queries";
import noop from "lodash/noop";
import { QueryEnvelope, QueryEnvelopeActionTypeName } from "./queryProcessing";
import { call, put } from "redux-saga/effects";
import { SagaIterator } from "redux-saga";

function envelopeWithPromise<TResult>(q: IDashboardQuery<TResult>): {
    promise: Promise<TResult>;
    envelope: QueryEnvelope;
} {
    const partialEnvelope = {
        onStart: noop,
        onSuccess: noop,
        onError: noop,
    };

    const promise = new Promise<TResult>((resolve, reject) => {
        partialEnvelope.onSuccess = resolve;
        partialEnvelope.onError = reject;
    });

    const envelope: QueryEnvelope = {
        type: QueryEnvelopeActionTypeName,
        query: q,
        ...partialEnvelope,
    };

    return {
        promise,
        envelope,
    };
}

/**
 * Runs the provided query and returns its result.
 *
 * @param q - query to run
 */
export function* query<TResult>(q: IDashboardQuery<TResult>): SagaIterator<TResult> {
    const { promise, envelope } = envelopeWithPromise(q);
    const waitForResult = (): Promise<TResult> => {
        return promise;
    };

    /*
     * dispatch the enveloped query. the envelope is setup so that upon query finish a promise will be resolved
     */
    yield put(envelope);

    return yield call(waitForResult);
}
