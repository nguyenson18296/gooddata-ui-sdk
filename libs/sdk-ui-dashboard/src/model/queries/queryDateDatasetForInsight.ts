// (C) 2021 GoodData Corporation
import { DashboardContext } from "../types/commonTypes";
import { SagaIterator } from "redux-saga";
import { ObjRef, serializeObjRef } from "@gooddata/sdk-model";
import { call } from "redux-saga/effects";
import { createCachedQueryService } from "../state/_infra/queryService";
import { PromiseFnReturnType } from "../types/sagas";
import { IDashboardQuery } from "./base";

/**
 * Given a reference to an insight, this query will obtain list of all date datasets that may be used
 * to filter it.
 *
 * @internal
 */
export interface QueryDateDatasetsForInsight extends IDashboardQuery<DateDatasetsForInsight> {
    type: "GDC.DASH/QUERY.INSIGHT.DATE.DATASETS";
    payload: {
        insightRef: ObjRef;
    };
}

/**
 * Creates action through which you can query dashboard component for information about the date data sets
 * that are applicable for the provided insight.
 *
 * @param insightRef - reference to insight
 * @internal
 */
export function queryDateDatasetsForInsight(insightRef: ObjRef): QueryDateDatasetsForInsight {
    return {
        type: "GDC.DASH/QUERY.INSIGHT.DATE.DATASETS",
        payload: {
            insightRef,
        },
    };
}

/**
 * @internal
 */
export type DateDatasetsForInsight = {
    data: any;
};

export const QueryDateDatasetsForInsightService = createCachedQueryService({
    queryName: "GDC.DASH/QUERY.INSIGHT.DATE.DATASETS",
    generator: dateDatasetsForInsightSaga,
    queryToCacheKey: (query: QueryDateDatasetsForInsight) => serializeObjRef(query.payload.insightRef),
});

//
// Query implementation
//

function loadDateDatasetsForInsight(
    _ctx: DashboardContext,
    _insightRef: ObjRef,
): Promise<DateDatasetsForInsight> {
    // this would normally do the necessary work
    return Promise.resolve({
        data: "some dummy data",
    });
}

function* dateDatasetsForInsightSaga(
    ctx: DashboardContext,
    query: QueryDateDatasetsForInsight,
): SagaIterator<DateDatasetsForInsight> {
    // this may do some more work then just a single call to async function...
    const result: PromiseFnReturnType<typeof loadDateDatasetsForInsight> = yield call(
        loadDateDatasetsForInsight,
        ctx,
        query.payload.insightRef,
    );

    return result;
}
