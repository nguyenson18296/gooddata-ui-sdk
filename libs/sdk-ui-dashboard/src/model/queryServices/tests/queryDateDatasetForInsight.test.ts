// (C) 2021 GoodData Corporation
import { DashboardTester, preloadedTesterFactory } from "../../tests/DashboardTester";
import { EmptyDashboardIdentifier, TestCorrelation } from "../../tests/Dashboard.fixtures";
import { idRef } from "@gooddata/sdk-model";
import { DateDatasetsForInsight, queryDateDatasetsForInsight } from "../../queries/insights";
import { createSliceNameForQueryCache } from "../../state/_infra/queryService";

describe("query date dataset for insight", () => {
    const cacheName = createSliceNameForQueryCache("GDC.DASH/QUERY.INSIGHT.DATE.DATASETS");
    let Tester: DashboardTester;
    beforeEach(preloadedTesterFactory((tester) => (Tester = tester), EmptyDashboardIdentifier, undefined));

    it("should create query cache in state", () => {
        expect(Tester.state()._queryCache[cacheName]).toBeDefined();
    });

    it("should do work when query action is dispatched", async () => {
        const result = await Tester.query<DateDatasetsForInsight>(
            queryDateDatasetsForInsight(idRef("blabla")),
        );

        expect(result.data).toEqual("some dummy data");
        expect(Tester.state()._queryCache[cacheName]).toMatchSnapshot();
    });

    it("should emit the right events", async () => {
        await Tester.query<DateDatasetsForInsight>(
            queryDateDatasetsForInsight(idRef("blabla"), TestCorrelation),
        );

        expect(Tester.emittedEventsDigest()).toMatchSnapshot();
    });
});
