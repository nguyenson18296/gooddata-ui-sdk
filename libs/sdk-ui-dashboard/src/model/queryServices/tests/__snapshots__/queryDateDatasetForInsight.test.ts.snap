// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query date dataset for insight should do work when query action is dispatched 1`] = `
Object {
  "entities": Object {
    "{\\"identifier\\":\\"blabla\\"}": Object {
      "query": Object {
        "correlationId": undefined,
        "payload": Object {
          "insightRef": Object {
            "identifier": "blabla",
          },
        },
        "type": "GDC.DASH/QUERY.INSIGHT.DATE.DATASETS",
      },
      "result": Object {
        "data": "some dummy data",
      },
      "status": "success",
    },
  },
  "ids": Array [
    "{\\"identifier\\":\\"blabla\\"}",
  ],
}
`;

exports[`query date dataset for insight should emit the right events 1`] = `
Array [
  Object {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.QUERY.STARTED",
  },
  Object {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.QUERY.COMPLETED",
  },
]
`;
